database/src/ops/blockchain.rs

```diff
@@ -1,9 +1,156 @@
+
+use cuprate_types::VerifiedBlockInformation;
+
+use crate::{
+    database::{DatabaseRo, DatabaseRw},
+    env::EnvInner,
+    error::RuntimeError,
+    ops::macros::doc_error,
+    tables::{
+        BlockBlobs, BlockHeights, BlockInfos, KeyImages, NumOutputs, Outputs, PrunableHashes,
+        PrunableTxBlobs, PrunedTxBlobs, RctOutputs, Tables, TablesMut, TxHeights, TxIds,
+        TxUnlockTime,
+    },
+    transaction::{TxRo, TxRw},
+    types::{BlockHash, BlockHeight, BlockInfo, KeyImage, Output, PreRctOutputId, RctOutput},
+};
 
 //---------------------------------------------------------------------------------------------------- Free Functions
-/// TODO
-pub fn height() {
-    todo!()
+/// Retrieve the height of the chain.
+///
+/// This returns the chain-tip, not the [`top_block_height`].
+///
+/// For example:
+/// - The blockchain has 0 blocks => this returns `0`
+/// - The blockchain has 1 block (height 0) => this returns `1`
+/// - The blockchain has 2 blocks (height 1) => this returns `2`
+///
+/// So the height of a new block would be `chain_height()`.
+#[doc = doc_error!()]
+#[inline]
+pub fn chain_height(
+    table_block_heights: &impl DatabaseRo<BlockHeights>,
+) -> Result<BlockHeight, RuntimeError> {
+    table_block_heights.len()
+}
+
+/// Retrieve the height of the top block.
+///
+/// This returns the height of the top block, not the [`chain_height`].
+///
+/// For example:
+/// - The blockchain has 0 blocks => this returns `Err(RuntimeError::KeyNotFound)`
+/// - The blockchain has 1 block (height 0) => this returns `Ok(0)`
+/// - The blockchain has 2 blocks (height 1) => this returns `Ok(1)`
+///
+/// Note that in cases where no blocks have been written to the
+/// database yet, an error is returned: `Err(RuntimeError::KeyNotFound)`.
+///
+#[doc = doc_error!()]
+#[inline]
+pub fn top_block_height(
+    table_block_heights: &impl DatabaseRo<BlockHeights>,
+) -> Result<BlockHeight, RuntimeError> {
+    match table_block_heights.len()? {
+        0 => Err(RuntimeError::KeyNotFound),
+        height => Ok(height - 1),
+    }
+}
+
+/// Check how many cumulative generated coins there have been until a certain [`BlockHeight`].
+///
+/// This returns the total amount of Monero generated up to `block_height`
+/// (including the block itself) in atomic units.
+///
+/// For example:
+/// - on the genesis block `0`, this returns the amount block `0` generated
+/// - on the next block `1`, this returns the amount block `0` and `1` generated
+///
+/// If no blocks have been added, this returns `0`.
+#[doc = doc_error!()]
+#[inline]
+pub fn cumulative_generated_coins(
+    block_height: &BlockHeight,
+    table_block_infos: &impl DatabaseRo<BlockInfos>,
+) -> Result<u64, RuntimeError> {
+    match table_block_infos.get(block_height) {
+        Ok(block_info) => Ok(block_info.cumulative_generated_coins),
+        Err(RuntimeError::KeyNotFound) if block_height == &0 => Ok(0),
+        Err(e) => Err(e),
+    }
+}
```

Comment:
...here.

Reply:
I don't think we need the if block_height here

Reply:
We need to special case `before block 0` but make sure this returns `Err` if the `block_height` doesn't exist, right?

If this returned `Ok(0)` on a height that didn't exist (other than before 0), the above call in `add_block()` would be proceeding with a bad value.